File: src/main/kotlin/magicbook/gtlitecore/api/worldgen/feature/AbstractFeature.java
package magicbook.gtlitecore.api.worldgen.feature;

import magicbook.gtlitecore.api.utils.GTLiteLog;
import magicbook.gtlitecore.api.utils.functions.TriConsumer;
import magicbook.gtlitecore.api.worldgen.AbstractWorldGenerator;
import net.minecraft.block.state.IBlockState;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraft.world.gen.NoiseGeneratorSimplex;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public abstract class AbstractFeature
{

    private NoiseGeneratorSimplex generatorSimplex;

    private final int seed;

    public final List<FeatureCondition> featureConditions = new ArrayList<>();
    public static final List<AbstractFeature> features = new ArrayList<>();

    protected AbstractWorldGenerator FEATURE_GROW_INSTANCE;
    protected AbstractWorldGenerator WORLD_GEN_INSTANCE;

    private int totalChunksChecked;
    private int totalChunksPlaced;

    public AbstractFeature(int seed)
    {
        this.seed = seed;
        features.add(this);
    }

    public abstract boolean generate(World worldIn,
                                     BlockPos.MutableBlockPos blockPos,
                                     Random random,
                                     TriConsumer<World, BlockPos, IBlockState> notifier);

    public void setWorld(World worldIn)
    {
        this.generatorSimplex = new NoiseGeneratorSimplex(
                new Random(worldIn.getSeed() + this.seed));
    }

    public double getRandomStrength(int chunkX, int chunkZ)
    {
        return this.generatorSimplex.getValue(
                chunkX * this.getPerlinScale(),
                chunkZ * this.getPerlinScale());
    }

    public double getPerlinScale()
    {
        return 0.04;
    }

    public void updatePlacePercentage(boolean disSucceed)
    {
        this.totalChunksChecked++;
        Double totalChunks = ((double) this.totalChunksPlaced / (this.totalChunksChecked / 100));
        if (disSucceed)
            this.totalChunksPlaced++;

        if (this.totalChunksChecked % 1000 == 0)
        {
            GTLiteLog.logger.info("Feature {} has been placed successfully"
                            + " in chunks {} percent of the time out of {} chunks checked",
                    this, totalChunks, this.totalChunksChecked);
        }
    }

    public AbstractWorldGenerator getWorldGenInstance()
    {
        return this.WORLD_GEN_INSTANCE;
    }

    public AbstractFeature addCondition(FeatureCondition condition)
    {
        this.featureConditions.add(condition);
        return this;
    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/api/worldgen/feature/AbstractFeature.kt
package gregtechlite.gtlitecore.api.worldgen.feature

import gregtechlite.gtlitecore.api.GTLiteLog
import gregtechlite.gtlitecore.api.worldgen.AbstractWorldGenerator
import net.minecraft.block.state.IBlockState
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraft.world.gen.NoiseGeneratorSimplex
import java.util.*

abstract class AbstractFeature(private val seed: Int)
{

    val featureConditions = mutableListOf<FeatureCondition>()

    // Instance for a worldgen feature.
    protected var featureGrowInstance: AbstractWorldGenerator? = null

    var worldGenInstance: AbstractWorldGenerator? = null
        protected set

    private var generatorSimplex: NoiseGeneratorSimplex? = null
    private var totalChunksChecked = 0
    private var totalChunksPlaced = 0
    open val perlinScale: Double
        get() = 0.04

    companion object
    {
        val features = ArrayList<AbstractFeature?>()
    }

    init
    {
        features.add(this)
    }

    abstract fun generate(worldIn: World?,
                          blockPos: BlockPos.MutableBlockPos?,
                          random: Random?,
                          notifier: (World?, BlockPos?, IBlockState?) -> Unit): Boolean

    fun setWorld(worldIn: World)
    {
        this.generatorSimplex = NoiseGeneratorSimplex(Random(worldIn.seed + this.seed))
    }

    fun getRandomStrength(chunkX: Int, chunkZ: Int): Double
    {
        return this.generatorSimplex!!.getValue(chunkX * this.perlinScale,
                                                chunkZ * this.perlinScale)
    }

    fun updatePlacePercentage(disSucceed: Boolean)
    {
        this.totalChunksChecked++
        val totalChunks = (this.totalChunksPlaced.toDouble() / (this.totalChunksChecked / 100))
        if (disSucceed) this.totalChunksPlaced++

        if (this.totalChunksChecked % 1000 == 0)
        {
            GTLiteLog.logger.info("Feature {} has been placed successfully"
                + " in chunks {} percent of the time out of {} chunks checked",
                this, totalChunks, this.totalChunksChecked)
        }
    }

    open fun addCondition(condition: FeatureCondition): AbstractFeature
    {
        this.featureConditions.add(condition)
        return this
    }

}
